### PR TITLE
Feature: Ability for users to trust third party certificates

### PR DESCRIPTION
**Pull request by BaconRanch**
sk89q uses his skmclauncher for easy, automatic distribution of client addons for his SKCraft multiplayer servers.  He does this by creating a ZIP archive containing all client addons and their corresponding configuration files.  In an update, the skmclauncher downloads and installs the contents of this ZIP archive to the end user's minecraft directory.  For security, however, this ZIP archive must be signed by one of three certificates that are included with the launcher itself: mclogin.cer, mcupdate.cer, or sk89q.cer. Since sk89q is the owner of sk89q.cer, he simply signs the ZIP archive with the sk89q.cer certificate.

As a server administrator myself, I too would like to take advantage of this automatic client update system.  Unfortunately, I do not have the ability to sign ZIP archives with any of the aforementioned certificates, and, as a result, any archive I create will not be installed by the launcher, as it fails to validate.  While I could create a fork of the launcher and include my own certificate, I decided it would be a better idea to include a new feature: the ability for users to trust a third party certificate.

Here's how it works.  In the custom update location's XML file, server administrators can add a URL that points to a certificate.  The launcher will download this certificate and it will present UI asking users if they want to trust it.  If a user says yes, the launcher will accept the certificate and it will allow any ZIP or JAR archives signed with the accepted certificate to validate for the corresponding update.  If, on the other hand, the user says no, the update is cancelled.

This feature takes a convenience once reserved only for sk89q and opens it up to any and all server administrators without compromising security.

By the way, I noticed that the current version of the skmclauncher simply denies the installation of file types it doesn't understand.  I've seen messages like "Not sure how to validate .txt files."  In response to this, I have also added the same feature for file extensions.  Server administrators can list file extensions in their update XML file, which the launcher will ask users to trust.  This is useful for various configuration files that, for one reason or another, a server admin may not choose to package up in a ZIP archive for distribution.

### CODE REVIEW

**Comment by md-5** [comment]
This looks very nice, I'll have to give it a double check before pulling

**Comment by BaconRanch** [comment]
Thanks!  There may be a bit of cleaning up to do.  The classes I added do not have an author attribute in the javadoc.  Writing `@author sk89q` just seemed a bit awkward, but for consistency, I understand if we need to add it.  I've added comments in the code that should explain everything, but let me know if you have any questions.  I'll be glad to answer them.

**Comment by sk89q** [comment]
This looks very nice. Something I sadly never found some time to add! I'll have to take a closer look too.

The `@author` attribute should refer to the original primary author(s), but it's extremely optional since we have Git to track changes anyway.

**Comment by BaconRanch** [comment]
Just thought of something... Currently, I have stored hashes of .cer files that the user trusts, as well as trusted file extensions, in a file called trust.xml, which is stored in the base directory of the configuration.  There are a couple reasons why this is a bad idea, and it would be much better to simply store these hashes and file extensions in the launcher's config.xml file.  I will make and commit this change in my fork within the next 24 hours.

File: src/main/java/com/sk89q/mclauncher/update/UpdateTrustManager.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            InputStream in = new BufferedInputStream(new FileInputStream(trustFile));
+            
+            Document doc = parseXml(in);
+            XPath xpath = XPathFactory.newInstance().newXPath();
+            
+            for(Node n : getNodes(doc, xpath.compile("/trust/certificates/certificate")))
+                localCertificateHashes.add(getValue(n).toLowerCase());
+            
+            for(Node n : getNodes(doc, xpath.compile("/trust/fileextensions/fileextension")))
+                localCertificateHashes.add(getValue(n).toLowerCase());
```
Comment (BaconRanch) [inline src/main/java/com/sk89q/mclauncher/update/UpdateTrustManager.java]: Whoops... this should have been `localFileExtensions.add(...`

**Comment by gknova61** [comment]
May i see an example XML utilizing the 2 features you have added to them?

**Comment by BaconRanch** [comment]
Sure!  I am assuming you wanted to see an example of the main custom update XML file:

``` xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<update>
    <latest>LATEST_UPDATE_NUM</latest>
    <packageurl>http://YOUR_SERVER/packages.xml</packageurl>
    <certificates>
        <certificate>http://YOUR_SERVER/certificate.cer</certificate>           
    </certificates>
    <fileextensions>
        <fileextension>xml</fileextension>
        <fileextension>cfg</fileextension>
        <fileextension>txt</fileextension>
    </fileextensions>
</update>
```

This example assumes you have placed your packages.xml file and any certificates in the root directory of your webserver, but you can of course place them anywhere.  In this example, I list only one certificate.  You can technically include as many certificates as necessary for all ZIP and JAR files that need to be validated, but I recommend signing all your ZIPs and JARs with a single certificate.  In fact, you could even package up all those ZIP and JAR client addons into a single ZIP archive and then sign that **one** archive with your certificate.

By the way, it is not necessary to include jar or zip extensions in the `<fileextensions>` portion of the xml file, as these file types are accepted by the launcher by default.

**Comment by gknova61** [comment]
Ah, that clears it up for me but the certificate.cer must be in the root directory of the web server thou. There can be no sub-folders as far as I can tell. This makes it so it can't be used with file hosting sites like Dropbox. If there is a sub-folder, the launcher will give an error saying that it can't write to the .minecraft folder.

**Comment by BaconRanch** [comment]
That was a bug.  I just committed a fix.

**Comment by md-5** [comment]
Just checking this does not break any existing functionality at all?

**Comment by BaconRanch** [comment]
Erm I have not necessarily tested all possibilities, but I made only minor modifications to the existing code.  Most of my code exists in the form of additions rather than modifications.

This new feature does modify some attributes in the config.xml file stored locally on the client and it also reads new attributes in the remote update xml file at a custom update location.  However, in both cases, it will work fine even if these new attributes are missing.  In the case of the local config.xml attributes missing, the launcher will add them in, even if they are blank.  In the case of the remote update xml file, if there are no attributes for certificates or file extensions, the launcher will simply proceed as normal and attempt to read the package file, proceeding with the update as it does in the current version.  In this case, my code for checking file extensions and certificates will, for the most part, be skipped.

So, from what I have tested, this new code does not break any functionality.  Running this new code and performing an update from the official Minecraft update source on the Default configuration, for example, works fine.

I recommend that you download and compile my commit and try out the update process on a couple of servers.  Again, from my tests, everything works as normal when no certificates or file extensions are found, so no current functionality should be broken.

**Comment by EntityReborn** [comment]
I can confirm, basic current functionality remains sane. I don't have a server with custom packages available, so I cannot test the custom url update scheme, nor the added functionality here. The code looks sane tho.

**Comment by sk89q** [comment]
Right now I want to hold off on merging in this PR because I rather have the launcher download sets of configurations for servers rather than having users manually add new configurations. The certificate would be bundled with the initial setup, and the user would only need to accept the certificate once. Essentially what a user would need to do is click a button, enter a URL, and click OK.

As for the file extensions list, I am not a fan of that solution. It undermines the reasoning for the signing, because it allows an attacker to install nearly any file anywhere on the file system. Originally my plan was to use XML-Dsig (http://en.wikipedia.org/wiki/XML_Signature) for all the XML files, but because of the implicit complexity of that (although Java does already have an easy-to-use implementation), I held off. I did not adapt the other code, which was already written, meaning that I did not add path checking and other security measures to make up for it. I did restrict files to only ones that could be verified though, which at least disallowed installing permanent files in arbitrary locations.

CertificateDownloader  also looks like code duplication that can be done away with.

**Comment by BaconRanch** [comment]
"Essentially what a user would need to do is click a button, enter a URL, and click OK."  

Currently, here are the required steps for creating a new configuration, starting with the "New Configuration" window:
1.) Type in a name for the configuration
2.) Optionally enter a custom local path (I think this should be mandatory- the default configuration should be the only one that uses the .minecraft folder, and we could automate the naming of the custom path by having the launcher create a folder with something like ".minecraft_GUID" where GUID is some unique identifier)
3.) Enter in the URL for the update xml file.
4.) Switch to the configuration and click "Play"

Both steps 1 and 2 could be done automatically.  Server admins could specify a Name in the update xml file, and the local folder could be ".minecraft_GUID" or something else if you think GUIDS are ugly.  Ultimately, this is a combination of UI changes and making good assumptions about Configuration and folder naming.  That can certainly be added if you like.

I had originally thought that a one time download of a certificate would be a good idea but, certificates eventually expire.  Even if the server admin sets an arbitrarily large expiration date, admins still may want to change certificates for one reason or another (lost private key, etc).  As a result, my code assumes that, for any given update, the certificates available at the time of the update should be able to verify all JARs and ZIPs downloaded in that update.  With this in mind, my code downloads the listed certificates each time an update takes place.  This way, the current certificates that the admin is using at the time of the update are used, and there are no problems with old, no longer used certificates.

As for the file extensions, I understand if a better implementation is in the works.  That was kind of an added feature that seemed to make sense at the time, but I'm sure there are more secure ways of allowing other file extensions.

CertificateDownloader does have some code duplication.  I had tried to avoid writing it altogether, as all it really does is read a collection of string URLs and download the associated files at those urls.  It then hashes these files.  I couldn't find any class with just that functionality, as as the Updater class, for example, also included tons of functionality related to working with the package xml file.  I did, however, remove a good amount of duplicate code by using your HttpDownloader instead of manually setting up an HttpURLConnection.

I'll make a new commit in a couple days that includes the use of HttpDownloader in the CertificateDownloader and also removes file extensions altogether.  I honestly believe that, with some modifications or additions, we can get this to work the way you want.

**Comment by EntityReborn** [comment]
A fix for the code duplication is you could move like code into an abstract class, and have Updater etc extend it. That said, there is a lot of stuff that could/should be abstracted, but would be a large task to pull off.

**Comment by BaconRanch** [comment]
I had originally thought about doing this. It will modify Updater heavily though, and the two really only share very little code.  I don't know if it's worth it.  If I were to do this, I would need to think about future ways in which the code could be modified.  Simply moving out a couple methods and keeping most of the functionality in the Updater class would make the abstract class only solve my particular problem.  If more updating / downloading classes need to be added later, that could make for lots of work.

**Comment by sk89q** [comment]
I currently do not have a solution for other file extensions. I've thought about it several times in the past, but the only solutions I've come across are to (1) sign the XML files, (2) require the files be in a signed container file (zip, jar), or (3) download separate signed checksum files. None of those solutions are pretty.

There is the Downloader interface now that abstracts all the download routines.

**Comment by gknova61** [comment]
@EntityReborn I run a small temporary server where this does work, and so does the added functionality. If you'd like to test it, please feel free to use this as a custom update URL for a test:
http://dl.dropbox.com/u/2813860/gkcraft.xml

**Comment by sk89q** [comment]
Err, I forgot to mention that option (4) is to use SSL, but I assume that that is off the table for most people. It's also unneeded overhead.

**Comment by BaconRanch** [comment]
I have modified `CertificateDownloader` to use the provided `HTTPDownloader` in `mclauncher.util` instead of manually setting up a Java `HTTPURLConnection`.  This removes a good amount of duplicate code.  What remains in `CertificateDownloader`'s `download()` method is now completely specific to downloading Certificates- so it belongs there.  I could remove the `addProgressListener()`, `removeProgressListener()`, and `fireStatusChange()` methods by extending the new `AbstractDownloader` as sk89q suggested.  Next, however, I plan to remove my File Extensions code.  I may tackle sk89q's File Extensions problem on my own, but chances are if he can't come up with a good solution, I certainly won't be able to.  Not sure what to do about that.

**Comment by sk89q** [comment]
Any progress on file extensions?

**Comment by BaconRanch** [comment]
Ok.  So File Extensions have been removed and the commit has been made.

I've started looking into the the new `Downloader` interface, as well as the `AbstractDownloader`, the `URLConnectionDownloader`, and the `SocketDownloader`. As I've discovered, all of these represent the new "re-implementation" of the low level `HttpDownloader`.

Like `Updater`, my `CertificateDownloader` consumes `HttpDownloader` and "pushes" events to the UI to display the status and progress of the download operation.  Both `CertificateDownloader` and `Updater` serve the same kind of responsibility: they are higher level objects that consume or "wrap around" the `HttpDownloader`.  Now, I have updated my `CertificateDownloader` to make use of the new `URLConnectionDownloader` (should I use `SocketDownloader` instead?) but I can't make any more code reductions on `CertificateDownloader` unless we create some sort of abstract base class specifically between the code that `Updater` and `CertificateDownloader` share.  As I stated earlier, these two classes share relatively little code (only the code that relates to UI notification) and creating a base class between them would offer, in my opinion, little value.  As a result, I think `CertificateDownloader` is pretty much complete.  Look over it yourself and see what you think.

By the way, what do you dislike about the security solution that forces all the files in an update to be in a signed ZIP or JAR container? I think having that requirement would be a good idea, as downloading a series of files in a large ZIP or JAR has the additional benefit that only one HTTP request would be sent to the custom update server, and fewer HTTP requests is a good thing.

**Comment by sk89q** [comment]
SocketDownloader was a test version to attempt to fix a problem where certain people would have the download constantly drop right in the middle. I rewrote it to be low-level so I could trace it down, but it didn't help. There isn't a really need to use it.

I don't have an issue with having a signed JAR for everything. That's how the launcher currently handles it. It's just not a solution many people would fall over loving I imagine.

**Comment by EntityReborn** [comment]
I have a hard time believing people wouldn't like signed JARs. Sure, more work for the dev, but end users will be able to trust that the code contains what the developer wanted, and that it's not some hacked version.

**Comment by sk89q** [comment]
I'd say that most end users have absolutely zero concern, at least, until something happens. It's just a 'turn off' thing for developers. I've heard people moan about using keytool in Java, which I find a bit misplaced, but it is definitely very daunting if you have no familiarity with public key cryptography.

**Comment by BaconRanch** [comment]
Provided that skmclauncher eventually gets a section on SK's Wiki (or some place for docs), all of this could be explained.  We could also possibly add some command line options to skmclauncher that developers and server admins could use to create a certificate and script out some test xml files for the custom update.  Just a thought.  Haven't even begun to think about trying to implement it, but at least with proper docs, most admins should be receptive to this solution.

**Comment by Elizafox** [comment]
If I may interject...

As somebody who wants to use this launcher for her own server, I have spoken to end-users about this (of varying technical skill). The consensus is that signed jars aren't really necessary and that if someone hacks the server you're publishing updates on, you have bigger problems.

Also, I see a few problems with this in regards to alternate XML files:
- If the XML file is permitted to change, so can be the certificate. Most users will just click accept for the certificate and pay it no mind. A truly malicious attacker can impersonate the person and say "yeah the old certificate was revoked" and then you have the whole problem of "most users don't really give a damn as long as it seems to work."
- Certificate compromise. If the old certificate was revoked due to compromise, how do you tell your savvy users "yeah it's ok to use this" without worrying about the aforementioned problem?
- Even if you sign the XML file (I personally would just use PGP but I doubt java has any easy to work with tools for that) how do you tell users the cert used to sign it is the canonical certificate? Users could easily be duped if they are new and there's been a compromise.

I am aware none of these problems are particularly new, but I think they are important issues that ought to be discussed before proceeding on whether or not signing the jars is worth it.

I know this probably seems to be biased against signing, but really I am neutral on the matter... I just want signing to be useful rather than create illusory security.

**Comment by sk89q** [comment]
I've considered that, hence the "most end users have absolutely no care," but it is already fully supported so it's not an issue. Generating a certificate can be made an easy to do task if a program is made to do it, so that's a non-issue.

A certificate change can be given with a very obvious alert window, so I don't think that is a problem. As for the first certificate that they have to verify, this is an issue with even SSH. You live with it or make plans to mitigate the issue.

**Comment by BaconRanch** [comment]
It's been a while since this PR was last updated and I wanted to hear about any current obstacles that might be preventing progress.  The real decision at this point is whether or not to go ahead with the packaging of all client mods into a single ZIP.  

Overall, however, determining the specifics of how server admins and players actually use this feature is really the final step and, unless we make major changes to what we've previously discussed, I think the code is pretty much complete.  Sure, there may be some UI changes to make some dialogs more obvious, but there's not much else I can think of.

What do you think, sk?

**Comment by ThomasGaubert** [comment]
Indeed, I'd also like to know the status of this PR.

@sk89q

**Comment by darkvip3r** [comment]
@BaconRanch  when i start the launcher it does not ask me if i would like to add the cert to a trust list, no option appears

**Comment by BaconRanch** [comment]
@darkvip3r Certificates are downloaded from a remote server during an update.  Users aren't given an option to add a certificate directly when the launcher starts.  Instead, a certificate is specified by a server admin in an XML file that the launcher reads.  Here's an example of what that file would look like:

``` xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<update>
    <latest>LATEST_UPDATE_NUM</latest>
    <packageurl>http://YOUR_SERVER/packages.xml</packageurl>
    <certificates>
        <certificate>http://YOUR_SERVER/certificate.cer</certificate>           
    </certificates>
</update>
```

There's no documentation to explain how this works yet, but users don't really need to supply anything.

**Comment by darkvip3r** [comment]
@BaconRanch  Thank you :) but how do i make a .cer file. i mean sorry for being a bit dumb. i don't understand this very well and there is not much to go on . but what do i use for the cert, if i cannot use what i got from your sklauncher repo. then what do i use to make a cert file ?

Sorry again

**Comment by sk89q** [comment]
I've been thinking about having a per-configuration option to disable the certificate check.

Simply because I doubt the documentation to explain how people can sign their .jars with a certificate will ever be made.

**Comment by darkvip3r** [comment]
@sk89q  yeah that would be awesome. and you are right i have looked around and found nothing on how to sign files :(

**Comment by BaconRanch** [comment]
I will try and add a certificate check disable feature sometime in the next week or two.  I've been busy with other things lately, but it shouldn't be much of a stretch.

I'll leave the certificate checking disabled by default on new configurations.

**Comment by BaconRanch** [comment]
I have looked at a disable feature and I realized that when the user creates a new configuration and selects "Use a custom update URL" it is implied that they wish to use third party certificates.  Custom update locations will almost always contain custom packages and these packages will likely have to be signed by third party certificates.  I was thinking we could simply change the user interface:

diag

Effectively, the "Use a custom update URL" checkbox enables both custom update locations and third party certificates, which makes sense, considering that 99% of custom update URLs will require a custom certificate.

I will be happy to commit this UI change to the PR if you agree.

By the way, we already have the certificate system working in such a way that when no custom update URL is specified, certificate checking doesn't happen.

**Comment by sk89q** [comment]
I ended up disabling the certificate check completely for custom URLs recently -- not sure if you noticed.

Either way, while something better would be preferred, I think it's more worthwhile at the moment to make the launcher's update even more flexible/powerful.

I was also thinking about replacing the manual XML save/load code with direct to/from XML serialization using a library I think that is bundled with Java. Adding more options is a chore when you have to manage the save file. (I also considered switching to like JSON, but we have to bundle another library, and it's still more work than automatic serialization.)

**Comment by BaconRanch** [comment]
Yes I actually just looked at your blog post and was about to edit my comment.  

Did you want me to update my fork to the latest version of the launcher and possibly consider this in the future or should we just discard this PR?  I am perfectly fine with either decision by the way.

**Comment by sk89q** [comment]
I have not looked at it closely enough to make a proper judgement, but realistically I don't think it will happen so I guess we'll have to put this on the back burner.

**Comment by darkvip3r** [comment]
@sk89q How do i edit the gui ? and also it would be awesome if you could pick a previous update, from a drop down. you know like technic do there

**Comment by sk89q** [comment]
Now irrelevant.

**Pull request closed without merging**